controllers/todo.js

```diff
@@ -5,36 +5,192 @@ const { ToDo, Token } = require("../models");
+  // which will also remove the collaborators from that TODO and also original creator of that TODO
+
+  ToDo.findOneAndReplace({_id:id,createdBy:user._id},{title:newTitle,createdBy:user._id},{new:true},function(err,updatedTodo) {
+    if(updatedTodo){
+    if(!err){res.status(200).json({"id":updatedTodo.id,
+                                   "title":updatedTodo.title})
+    }else
+    { 
+      console.log(err)
+      res.status(400).send("Error in updating todo.")} } 
+    else res.status(400).send("Error in updating todo.")
+  })
+
+};      
+
+
+                                                                                                       
 const editToDoPatch = async (req, res) => {
   // Change the title of the Todo with given id, and get the new title as response
+  const {id}=req.params
+  const user=req.user
+  ToDo.findOneAndUpdate({$or: [{_id:id,createdBy:user._id},{_id:id,collaborators:user._id}]},{$set:req.body},{new:true},function(err,updatedTodo){
+    if(!err && updatedTodo){res.status(200).json({"id":updatedTodo.id,
+                                   "title":updatedTodo.title})
+    }else
+    { res.status(400).send("Error in updating todo.")}  
+  })
 };
 
+
+
 const deleteToDo = async (req, res) => {
   //  Delete the todo with given id
+  
+  const {id}=req.params
+  const user = req.user
+  ToDo.findOne({$or: [{_id:id,createdBy:user._id},{_id:id,collaborators:user._id}]},function (err, todo){
+
+    if(todo){
+      ToDo.deleteOne({_id:id},function(err){
+        if(!err) res.status(200).send("Deleted the dodo successfully.")
+        else { res.status(400).send("Error in deleting the todo.")}
+      })
+    }else{
+      res.status(400).send("Wrong TODO id or you do not have the permission to delete the TODO.")
+    }
+  })
 };
 
+
+const addCollaborator = async (req, res) => {
+  const { id } = req.params
+  const { userid } = req.body
+  const user = req.user
+  ToDo.findOne({_id:id,createdBy:user._id},function (err, todo){
+    if(!err && todo){
+
+    if(todo.collaborators.includes(userid)){return res.status(200).send("This user is already a collaborator for the TODO.")}
+    todo.collaborators.push(userid)
+    todo.save()
+    res.status(200).json({"Message":"User with the given id successfully added as collaborator.",
+                          "TodoId":todo._id,
+                          "TodoTitle":todo.title})
+  }else{
+    res.status(400).send("Error in adding collaborator.")
+  }
+
+  })
+
+}
+
+const removeCollaborator = async (req, res) => {
+  const { id } = req.params
+  const { userid }=req.body
+  
+  const user = req.user
+  ToDo.findOne({_id:id,createdBy:user._id},function (err, todo){
+    if(!err && todo){
+
+    if(!todo.collaborators.includes(userid)){return res.status(200).send("The user you are tying to remove is already not a collaborator.")}
+    ToDo.findOneAndUpdate({_id:id,createdBy:user._id},{$pull:{collaborators:userid}},function(err,updatedTodo){
```

Comment:
Any specific reason as to why using `$pull` here but  `.push` in the addition of collaborators?